File: unpackerGo/filesystem.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// FileSystem represents the game filesystem with GPK packages
type FileSystem struct {
	root string
	gpks []*GPK
}

// NewFileSystem creates a new filesystem instance
func NewFileSystem(gameRoot string) (*FileSystem, error) {
	fs := &FileSystem{
		root: gameRoot,
		gpks: make([]*GPK, 0),
	}

	err := fs.findArchives()
	if err != nil {
		return nil, fmt.Errorf("failed to find archives: %w", err)
	}

	return fs, nil
}

// findArchives locates and mounts all GPK files in the packs directory
func (fs *FileSystem) findArchives() error {
	packsRoot := filepath.Join(fs.root, "packs")

	err := filepath.Walk(packsRoot, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		if !info.IsDir() && strings.ToUpper(filepath.Ext(info.Name())) == ".GPK" {
			err := fs.mountGPK(path)
			if err != nil {
				fmt.Printf("Warning: Failed to mount GPK %s: %v\n", path, err)
			} else {
				fmt.Printf("Mounted package: %s\n", info.Name())
			}
		}
		return nil
	})

	return err
}

// mountGPK loads a GPK file and adds it to the filesystem
func (fs *FileSystem) mountGPK(fileName string) error {
	gpk := NewGPK()
	err := gpk.Load(fileName)
	if err != nil {
		return err
	}

	fs.gpks = append(fs.gpks, gpk)
	return nil
}

// UnpackAll unpacks all mounted GPK files
func (fs *FileSystem) UnpackAll() error {
	for _, gpk := range fs.gpks {
		fmt.Printf("Unpacking: %s\n", gpk.GetName())
		outputDir := filepath.Join(fs.root, gpk.GetName())
		err := gpk.UnpackAll(outputDir)
		if err != nil {
			return fmt.Errorf("failed to unpack %s: %w", gpk.GetName(), err)
		}
	}
	return nil
}

// Open opens a file from the filesystem (either from disk or GPK)
func (fs *FileSystem) Open(filename string) (*GPKFile, error) {
	// First check if file exists on disk
	fullPath := filepath.Join(fs.root, filename)
	if _, err := os.Stat(fullPath); err == nil {
		return NewGPKFileFromDisk(fullPath)
	}

	// Extract package name and file path
	parts := strings.SplitN(filename, "/", 2)
	if len(parts) != 2 {
		return nil, fmt.Errorf("invalid filename format: %s", filename)
	}

	pkg := parts[0]
	filePath := parts[1]
	normalizedName := normalizeName(pkg, filePath)

	// Search in GPK packages
	for _, gpk := range fs.gpks {
		if strings.EqualFold(gpk.GetName(), pkg) {
			return gpk.Open(normalizedName)
		}
	}

	return nil, fmt.Errorf("file not found: %s", filename)
}

// List lists files matching a pattern in a specific package
func (fs *FileSystem) List(mask string) ([]string, error) {
	parts := strings.SplitN(mask, string(filepath.Separator), 2)
	if len(parts) != 2 {
		return nil, fmt.Errorf("invalid mask format: %s", mask)
	}

	pkg := parts[0]
	pattern := parts[1]

	for _, gpk := range fs.gpks {
		if strings.EqualFold(gpk.GetName(), pkg) {
			return gpk.List(pattern), nil
		}
	}

	return []string{}, nil
}

// GetRoot returns the root directory of the filesystem
func (fs *FileSystem) GetRoot() string {
	return fs.root
}

// normalizeName normalizes file names based on package type
func normalizeName(pkg, name string) string {
	pkgUpper := strings.ToUpper(pkg)

	if strings.HasPrefix(pkgUpper, "SYSSE") || strings.HasPrefix(pkgUpper, "SE") || strings.HasPrefix(pkgUpper, "VOICE") {
		return name + ".ogg"
	} else if strings.HasPrefix(pkgUpper, "BGM") {
		return name + "_loop.ogg"
	} else if strings.HasPrefix(pkgUpper, "EVENT") {
		return name + ".PNG"
	}

	return name
}

// NormalizeName normalizes a file name using the filesystem's package information
func (fs *FileSystem) NormalizeName(name string) string {
	parts := strings.SplitN(name, "/", 2)
	if len(parts) != 2 {
		return name
	}
	return normalizeName(parts[0], parts[1])
}

File: unpacker_cli/filesystem.cpp
#include "filesystem.h"
#include <iostream>
#include <filesystem>
#include <algorithm>

#ifdef _WIN32
    const char PATH_SEPARATOR = '\\';
    const std::string PATH_SEPARATOR_STR = "\\";
#else
    const char PATH_SEPARATOR = '/';
    const std::string PATH_SEPARATOR_STR = "/";
#endif

FileSystem::FileSystem(const std::string& gameRoot) : root(gameRoot) {
    // Ensure the path ends with a separator
    if (!root.empty() && root.back() != PATH_SEPARATOR) {
        root += PATH_SEPARATOR_STR;
    }
    findArchives();
}

FileSystem::~FileSystem() {
    for (GPK* gpk : gpks) {
        delete gpk;
    }
}

void FileSystem::unpack_all() {
    for (GPK* gpk : gpks) {
        std::cout << "Unpacking: " << gpk->getName() << std::endl;
        std::string outputDir = root + gpk->getName() + PATH_SEPARATOR_STR;
        gpk->unpack_all(outputDir);
    }
}

void FileSystem::findArchives() {
    std::string packsRoot = root + "packs" + PATH_SEPARATOR_STR;
    
    try {
        if (!std::filesystem::exists(packsRoot)) {
            std::cout << "Warning: packs directory not found at: " << packsRoot << std::endl;
            return;
        }
        
        for (const auto& entry : std::filesystem::directory_iterator(packsRoot)) {
            if (entry.is_regular_file()) {
                std::string filename = entry.path().filename().string();
                std::string extension = entry.path().extension().string();
                
                // Convert to uppercase for comparison
                std::transform(extension.begin(), extension.end(), extension.begin(), ::toupper);
                
                if (extension == ".GPK") {
                    std::string fullPath = entry.path().string();
                    mountGPK(fullPath);
                    std::cout << "Mounted package: " << filename << std::endl;
                }
            }
        }
    } catch (const std::filesystem::filesystem_error& ex) {
        std::cerr << "Filesystem error: " << ex.what() << std::endl;
    }
}

void FileSystem::mountGPK(const std::string& fileName) {
    GPK* gpk = new GPK();
    if (gpk->load(fileName)) {
        gpks.push_back(gpk);
    } else {
        delete gpk;
    }
}

std::string FileSystem::normalize_name(const std::string& name) {
    size_t slashPos = name.find('/');
    if (slashPos != std::string::npos) {
        std::string pkg = name.substr(0, slashPos);
        return normalize_name(pkg, name);
    }
    return name;
}

std::string FileSystem::normalize_name(const std::string& pkg, const std::string& name) {
    if (pkg.substr(0, 5) == "SysSe" || pkg.substr(0, 2) == "Se" || pkg.substr(0, 5) == "Voice") {
        return name + ".ogg";
    } else if (pkg.substr(0, 3) == "BGM") {
        return name + "_loop.ogg";
    } else if (pkg.substr(0, 5) == "Event") {
        return name + ".PNG";
    } else {
        return name;
    }
}
